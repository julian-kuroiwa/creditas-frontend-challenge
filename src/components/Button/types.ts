import { ButtonHTMLAttributes, ReactNode } from 'react'

export type ButtonStyleProps = {
  size?: 'small' | 'medium' | 'large'
}

export interface ButtonProps
  extends ButtonStyleProps,
    ButtonHTMLAttributes<HTMLButtonElement> {
  children: ReactNode
  isLoading?: boolean
}
