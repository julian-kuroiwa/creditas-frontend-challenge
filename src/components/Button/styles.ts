import styled from 'styled-components'

import { theme } from '~/styles/theme'
import { ButtonStyleProps } from './types'

export const Button = styled.button<ButtonStyleProps>`
  border: none;
  border-radius: 16px;
  font-size: ${theme.font.text.size.m};
  min-height: 32px;
  max-height: 60px;

  transition: all 0.2s;
  width: 100%;
  background-color: ${theme.colors.button.default};

  &:hover {
    box-shadow:
      rgba(0, 0, 0, 0.16) 0px 16px 12px -8px,
      rgba(0, 0, 0, 0.12) 0px 16px 8px -12px;
  }

  &[disabled] {
    background-color: ${theme.colors.gray.medium};
    cursor: default;
    display: flex;
    justify-content: center;
    align-items: center;

    &:hover {
      box-shadow: none;
    }
  }

  ${({ size }) => {
    switch (size) {
      case 'small':
        return `
          border-radius: 8px;
          height: 32px;
          font-size: ${theme.font.text.size.s};
        `
      case 'medium':
        return `height: 44px;`
      case 'large':
        return `
          height: 60px;
        `
    }
  }}
`
