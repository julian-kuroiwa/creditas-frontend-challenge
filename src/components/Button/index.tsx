import { ButtonProps } from './types'
import * as S from './styles'
import { Loader } from '../Loader'

const Button = ({
  size = 'medium',
  children,
  isLoading,
  ...props
}: ButtonProps) => {
  return (
    <S.Button size={size} {...props} disabled={isLoading}>
      {isLoading ? <Loader data-testid="loader-button" /> : children}
    </S.Button>
  )
}

export default Button
