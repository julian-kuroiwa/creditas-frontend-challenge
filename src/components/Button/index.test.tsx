import { fireEvent, render } from '@testing-library/react'
import { vi } from 'vitest'

import Button from './'

const onClick = vi.fn()
const buttonLabel = 'Click me'

describe('Button', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(
      <Button onClick={onClick}>{buttonLabel}</Button>,
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should fire onClick method if the button is clicked', () => {
    const { getByRole } = render(
      <Button onClick={onClick}>{buttonLabel}</Button>,
    )

    const button = getByRole('button')
    fireEvent.click(button)

    expect(onClick).toHaveBeenCalled()
  })

  it('should have a loading if isLoading is passed', () => {
    const { getByTestId } = render(
      <Button onClick={onClick} isLoading>
        {buttonLabel}
      </Button>,
    )

    const loader = getByTestId('loader-button')

    expect(loader).toBeInTheDocument()
  })
})
