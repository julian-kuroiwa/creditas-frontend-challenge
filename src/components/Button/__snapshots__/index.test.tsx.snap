// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > should match snapshot 1`] = `
<DocumentFragment>
  <button
    class="sc-FEMpB blPeKa"
  >
    Click me
  </button>
</DocumentFragment>
`;
